scalanet/src/io/iohk/scalanet/peergroup/UDPPeerGroup.scala

```diff
@@ -34,18 +43,31 @@ import scala.util.control.NonFatal
   * @param codec a decco codec for reading writing messages to NIO ByteBuffer.
   * @tparam M the message type.
   */
-class UDPPeerGroup[M](val config: Config)(
+class UDPPeerGroup[M](val config: Config, cleanupScheduler: Scheduler = Scheduler.singleThread("cleanup-thread"))(
     implicit codec: Codec[M],
     bufferInstantiator: BufferInstantiator[ByteBuffer],
     scheduler: Scheduler
 ) extends TerminalPeerGroup[InetMultiAddress, M]() {
 
   private val log = LoggerFactory.getLogger(getClass)
+
   val serverSubject = ConnectableSubject[ServerEvent[InetMultiAddress, M]]()
 
   private val workerGroup = new NioEventLoopGroup()
 
-  private val activeChannels = new ConcurrentHashMap[ChannelId, ChannelImpl]().asScala
+  private[peergroup] val activeChannels = new ConcurrentHashMap[ChannelId, ChannelImpl]()
+
+  private def cleanupTask = {
```

Comment:
I think maybe `Scheduler.scheduleWithFixedDelay` would be better. In any case, are you certain that `cleanup.cancel()` actually stops the cleanup thread? I'm not sure, but it seems more likely that `cancel()` will stop the `foreach` callback but not stop the underlying `Observable`. Plus, there's no test to verify that `onComplete` is invoked on the val returned by `Observable.intervalAtFixedRate`
You could tidy this up a bit. There is a `def`, only used in the line below but if somebody did call it from somewhere else, it would create a second cleanup schedule. Maybe

Reply:
I think you are right, `Scheduler.scheduleWithFixedDelay` will be a better choice here. I did not find satisfactory explanation of `Observable.intervalAtFixedRate` behaviour when cancelled, and scheduler docs are much clearer about it.